### DISCUSSION TITLE
Improved handling of users and role collections

### DISCUSSION THREAD

**Discussion by rui1610** [SAP-archive/btp-setup-automator#119] [Announcements]
With the recent [pull request 109](https://github.com/SAP-samples/btp-setup-automator/pull/109) we have introduced an improved handling of users and role collections.

Now there is a **clear separation** from the actual use case specific information (services and role collections defined in the usecase.json file) to the actual users (usergroups defined in the parameters.json file).

## Changes for the parameters.json file

Within the parameters.json file you can add now a new attribute called `myusergroups`.  Each entry in that attribute has

- a `name`(to be used as a reference in the usecase.json file)
- a list of `members` with their email addresses

Here's an example (checkout the [default parameters.json file](https://github.com/SAP-samples/btp-setup-automator/blob/main/parameters.json):
```json
...
  "myusergroups": [
    {
      "name": "admins",
      "members": [
        "john.admin@test.com",
        "jane.admin@test.com"
      ]
    },
    {
      "name": "developers",
      "members": [
        "tony@stark.com"
      ]
    }
]
...
```

### Assigning users to a role collection of a service

Within the usecase.json file you can now assign a required role collection of an app subscription to a usergroup you've defined in the parameter.json file (as described above). With that the parameters.json file stays clean and stable. If one of the usergroups listed under `assignedUserGroupsFromParameterFile` is not found in the parameters.json file, it will simply assign the user, that is running the script to the corresponding role collection.

Here's an example (checkout the [default usecase.json file](https://github.com/SAP-samples/btp-setup-automator/blob/main/usecases/released/cap_app_launchpad.json):

```json
...
    {
      "category": "APPLICATION",
      "name": "SAPLaunchpad",
      "plan": "standard",
      "requiredrolecollections": [
        {
          "name": "Launchpad_Admin",
          "assignedUserGroupsFromParameterFile": [
            "developers"
          ]
        }
      ]
    }
...
```
### Assigning users to a role collection

Within the usecase.json file you can now as well assign users to existing role collections or even define role collections, that the btp-setup-automator will create for you in the BTP in the newly introduced attribute `assignrolecollections `.

Here's an example (checkout the [default usecase.json file](https://github.com/SAP-samples/btp-setup-automator/blob/main/usecases/released/cap_app_launchpad.json):
```json
...
  "assignrolecollections": [
    {
      "name": "Global Account Administrator",
      "type": "account",
      "level": "global account",
      "assignedUserGroupsFromParameterFile": [
        "admins"
      ]
    },
...
```
In that example the usergroup `admins` (defined in a [parameters.json file](https://github.com/SAP-samples/btp-setup-automator/blob/main/parameters.json) is assigned to the role collection `Global Account Administrator`. The attribute `type` and `level` define where this role collection exists.

**Comment by dachtera** [comment]
Is there information documented anywhere on how to create a new role collection group, and assign it other roles?  We would like to consolidate multiple groups for a team into one role collection to help streamline both the setup and maintenance process reguarding permissions and access; 

Same group 123 created, but we want to add Launchpad Admin, Cloud Admin, etc to the '123' role, instead of having to specify each roll group to each user.
![Screen Shot 2022-10-18 at 2 16 21 PM](https://user-images.githubusercontent.com/115640072/196523439-64b0efb1-fe65-4e68-93c6-6bce2925ac7e.png)

**Reply by lechnerc77** [reply]
@dachtera : The assignment of custom role collections is possible as described above. using **existing** role collections
The **creation** of custom role collections **and assignment** of roles and users via the btp-setup-automator is not possible as of today. 

Please open a [feature request](https://github.com/SAP-samples/btp-setup-automator/issues/new?assignees=&labels=enhancement&template=feature-request.yml&title=%5BFEATURE+REQUEST%5D+%3Ctitle%3E).

**Reply by lechnerc77** [reply]
Feature request is open <https://github.com/SAP-samples/btp-setup-automator/issues/268>

**Reply by lechnerc77** [reply]
Functionality provided via PR #273
